### ISSUE TITLE
querying `t_blocks`

### ISSUE THREAD

**Issue by sivathanuk** [wealdtech/chaind#68]
I have synced both the nodes,

execution node: Nethermind
Nethermind config:
```YAML
sudo ./Nethermind.Runner \
--config mainnet \
--JsonRpc.Enabled true \
--JsonRpc.JwtSecretFile /keystore/jwtsecret.hex \
--baseDbPath /nethermind_db/mainnet \
--Metrics.Enabled true \
--HealthChecks.Enabled true
```
consenus node: Lighthouse
Lighthouse config:
```YAML
sudo ./lighthouse bn \
--network mainnet \
--execution-endpoint http://localhost:8551 \
--datadir /data-mainnet \
--execution-jwt /keystore/jwtsecret.hex \
--checkpoint-sync-url https://beaconstate.ethstaker.cc \
--disable-deposit-contract-sync \
--metrics  \
--debug-level info  \
--eth1 \
--http \
--http-allow-sync-stalled \
--enr-udp-port=9000 \
--enr-tcp-port=9000 \
--discovery-port=9000
```
---
ans set up chaind. Here is my config for chaind,

```YAML
sudo ./chaind \
--chaindb.url=postgres://chain:secret@127.0.0.1:5432 \
--chaindb.max-connections=16 \
--eth2client.address=127.0.0.1:5052 \
--eth1client.address=127.0.0.1:8545 \
--log-level=trace \
--blocks.enable=true \
--beacon-committees.enable=true \
--proposer-duties.enable=true \
--finalizer.enable=true
```
> note: I am using port 5052 instead of port 5051 as recommended in the [lighthouse docs](https://lighthouse-book.sigmaprime.io/api-bn.html#starting-the-server) 
---
I just ran `chaind`. `chaind` was launched successfully and running, showing me the below list of tables for `\dt` command:
```shell
chain=# \dt
 public | t_attestations              | table | chain
 public | t_attester_slashings        | table | chain
 public | t_beacon_committees         | table | chain
 public | t_block_execution_payloads  | table | chain
 public | t_block_summaries           | table | chain
 public | t_blocks                    | table | chain
 public | t_chain_spec                | table | chain
 public | t_deposits                  | table | chain
 public | t_epoch_summaries           | table | chain
 public | t_eth1_deposits             | table | chain
 public | t_fork_schedule             | table | chain
 public | t_genesis                   | table | chain
 public | t_metadata                  | table | chain
 public | t_proposer_duties           | table | chain
 public | t_proposer_slashings        | table | chain
 public | t_sync_aggregates           | table | chain
 public | t_sync_committees           | table | chain
 public | t_validator_balances        | table | chain
 public | t_validator_epoch_summaries | table | chain
 public | t_validators                | table | chain
 public | t_voluntary_exits           | table | chain
```

However when I query the table `t_blocks` I get only 32 rows
```SQL
chain=# select * from t_blocks;
```
Also, the below query fetched 31 as the output 
```sql
chain=# select max(f_slot) from t_blocks;
# output:   31
```

Also, the below tables are empty,
- `t_block_summaries`
- `t_epoch_summaries`
- `t_proposer_slashings`
- `t_deposits`
- `t_validator_epoch_summaries`

`chaind` is still running. Just checking in, to understand if I made any mistake in the `chaind` config.

 Can you please guide to fix this and re-populate `t_blocks` table (with all slots) from the node? Kindly let me know if you need any additional information.

**Comment by mcdee** [comment]
The first question is if lighthouse has synced fully.  See the section "Backfilling blocks" at https://lighthouse-book.sigmaprime.io/checkpoint-sync.html#backfilling-blocks and confirm there that the historical sync is complete.

You will also need some level of states to be reconstructed.  https://lighthouse-book.sigmaprime.io/checkpoint-sync.html?highlight=archive#reconstructing-states contains information about this.

Once both of these are in place `chaind` should be able to sync up blocks.  If it does not appear to be syncing blocks then the first place to check is the `chaind` logs to see if there are any issues being reported there.

Items such as summaries do not, in general, appear until after the primary data has been synced.  As such, until the issue with filling in `t_blocks` is addressed you shouldn't worry about the other tables.

**Comment by sivathanuk** [comment]
Thanks a lot for the quick response. I have checked ,
- `Backfilling blocks` : confirmed that historical sync is complete (could be because I ran `fast sync` and probably no backfilling is required)
- `states to be reconstructed` : Is this check relevant for `fast-sync` as well? 
	- I have run nethermind fast-sync (due to hardware limitations: `archive sync`  requires ~12TB memory,  default `fast-sync` requires ~500GB)
	- From [nethermind doc](https://docs.nethermind.io/nethermind/ethereum-client/sync-modes),
	   > "fast sync: downloads only the latest state, headers, and optionally bodies and receipts"
	- list of sync modes that nethermind offers: [network config](https://docs.nethermind.io/nethermind/ethereum-client/configuration/network)
---
**Queries**:
1. Do I need to re-run nethermind node in `archive sync` mode ?
2. Is there a way I can use `chaind` to query data from the `fast-synced` node?
	- If `chaind` can be used to query data from the `fast-synced` node, what tables does `chaind` populate?
---
**Note**:  I am able to fetch historic slot information (checked random slot ids) using the below queries,
```shell
curl -X GET "http://localhost:5052/eth/v1/beacon/headers/50" -H  "accept: application/json" | jq
```
output :
![image](https://user-images.githubusercontent.com/62845445/210200430-062a6661-fcba-436c-8765-0ac13847cc7b.png)

```shell
curl -X GET "http://localhost:5052/eth/v1/beacon/headers/75000" -H  "accept: application/json" | jq
```
output:
![image](https://user-images.githubusercontent.com/62845445/210200537-8f02d1b2-6c95-45df-9758-1b9efd513d08.png)



```shell
curl -X GET "http://localhost:5052/eth/v1/beacon/headers/900002" -H  "accept: application/json" | jq
```
![image](https://user-images.githubusercontent.com/62845445/210200555-7b2d154d-0798-43dd-8c74-aa9336c60d35.png)


recent slot
```shell
curl -X GET "http://localhost:5052/eth/v1/beacon/headers/5484702" -H  "accept: application/json" | jq
```
output:
![image](https://user-images.githubusercontent.com/62845445/210200571-a5bba1a8-6b3c-4e41-ab00-0e360623ded3.png)

**Comment by mcdee** [comment]
You do not need the execution client to be synced in archive mode for `chaind`, as `chaind` does not access the additional archive data on the execution node.

The historic information you posted shows `execution_optimistic` to be `true`, which suggests that the node is not yet fully synced.  As it has been a few days, are there any changes?

You can also change the `chaind` log level to `trace`, where the logs will give a lot more information about what each individual process is doing.

**Comment by sivathanuk** [comment]
Thanks a lot for the response. I have fixed the db issues with the sync data and re-run the nodes. The nodes are now synced.  I'll  change the log level to trace, re-run chaind and share the updates. 

---

##### Nethermind:
<img width="1139" alt="image" src="https://user-images.githubusercontent.com/62845445/212255280-163e82aa-3798-444c-8198-12e81a26ab92.png">


##### Lighthouse:
<img width="1142" alt="image" src="https://user-images.githubusercontent.com/62845445/212255370-83608ead-a74f-4e6d-9643-30b51059c34e.png">

**Comment by sivathanuk** [comment]
Thanks a lot for the response. I have re-run the clients and the node is fully synced now,

#### Nethermind:
<img width="1162" alt="image" src="https://user-images.githubusercontent.com/62845445/212409344-edfd051c-2f1c-4fc6-a6fb-17fb839beef4.png">

#### Lighthouse:
<img width="1141" alt="image" src="https://user-images.githubusercontent.com/62845445/212412299-9f7811eb-ae1a-4797-8e49-7a9f63e8380b.png">

---
#### Chaind
1. I have set log level to trace and started running `chaind`. `chaind` is running  without throwing any error message (and the log info runs very rapidly)
2. I have confirmed again that the node is fully synced
3. However, when I query the tables, I can see that the tables are not completely populated
4. I have shared the output from some of the queries below. Not sure, if I'm missing something, can you please guide me? Please let me know if you need any additional log info
---
#### Output from queries
1. `t_blocks` table has only 26 rows,
```sql
chain=# select count(*) from t_blocks;
 count 
-------
    26
(1 row)

chain=# select max(f_slot) from t_blocks;
 max 
-----
  31
(1 row)
```
2. When I queried the `t_metadata` table, I got the below output,
<img width="1501" alt="image" src="https://user-images.githubusercontent.com/62845445/212410864-a669a5ae-388e-4325-b7d9-f316ade82b44.png">

3. `count(*)` output from few other tables

```sql
chain=# select count(*) from t_attestations;
   276

chain=# select count(*) from t_beacon_committees; 
   160

chain=# select count(*) from t_block_execution_payloads; 
     0

chain=# select count(*) from t_block_summaries; 
     0

chain=# select count(*) from t_chain_spec; 
    92

chain=# select count(*) from t_validators; 
 500294

chain=# select count(*) from t_sync_aggregates; 
     0

chain=# select count(*) from t_fork_schedule; 
     3

chain=# select * from t_fork_schedule; 
       0 | \x00000000
   74240 | \x01000000
  144896 | \x02000000

```

**Comment by mcdee** [comment]
The summarizer metadata looks wrong, I'll look in to that, but the rest looks okay.  Could you grep the logs for `"module": "blocks"` to see what is happening with populating blocks, as pretty much everything else requires that information to be populated before it will operate.

**Comment by sivathanuk** [comment]
Sure, thanks a lot. Just adding my `chaind` config for more context. Here is my `chaind` config,

```shell
sudo ./chaind \
--chaindb.url=postgres://chain:secret@127.0.0.1:5432 \
--chaindb.max-connections=16 \
--eth2client.address=127.0.0.1:5052 \
--eth1client.address=127.0.0.1:8545 \
--log-level=trace \
--blocks.enable=true \
--beacon-committees.enable=true \
--proposer-duties.enable=true \
--finalizer.enable=true
```

To grep the logs for "module": "blocks",  I added the log file `--log-file=/var/log/chaind.log` to the config below. `chaind` doesn't run and there is no log info (no error msg as well) 

```shell
sudo ./chaind \
--chaindb.url=postgres://chain:chaind@127.0.0.1:5432 \
--chaindb.max-connections=16 \
--eth2client.address=127.0.0.1:5052 \
--eth1client.address=127.0.0.1:8545 \
--log-level=trace \
--log-file=/var/log/chaind.log \
--blocks.enable=true \
--beacon-committees.enable=true \
--proposer-duties.enable=true \
--finalizer.enable=true
```
Do I need to do any modification to the config  to store the logs?

**Comment by sivathanuk** [comment]
Sure, thanks a lot. Just adding my `chaind` config for more context. Here is my `chaind` config,

```YAML
sudo ./chaind \
--chaindb.url=postgres://chain:secret@127.0.0.1:5432 \
--chaindb.max-connections=16 \
--eth2client.address=127.0.0.1:5052 \
--eth1client.address=127.0.0.1:8545 \
--log-level=trace \
--blocks.enable=true \
--beacon-committees.enable=true \
--proposer-duties.enable=true \
--finalizer.enable=true
```

To grep the logs for "module": "blocks", I added the log file `--log-file=/var/log/chaind.log` to the config below. `chaind` doesn't run and there is no log info (no error msg as well)

```YAML
sudo ./chaind \
--chaindb.url=postgres://chain:secret@127.0.0.1:5432 \
--chaindb.max-connections=16 \
--eth2client.address=127.0.0.1:5052 \
--eth1client.address=127.0.0.1:8545 \
--log-level=trace \
--log-file=/var/log/chaind.log \
--blocks.enable=true \
--beacon-committees.enable=true \
--proposer-duties.enable=true \
--finalizer.enable=true
```
Do I need to do any  modification to the config to store the logs?

**Comment by mcdee** [comment]
If the first command works then the second should, if the only change is to add `--log-file`.  Or you could use the first command and add `2>&1 | grep '"module": "blocks"` at the end of the it.

**Comment by sivathanuk** [comment]
Got it, found the file in `/var/log/`.  Didn't get any matches for `grep '"module": "blocks"' chaind.log`

**Comment by mcdee** [comment]
Sorry, wrong string.  Try `grep '"service":"blocks"' chaind.log`

**Comment by sivathanuk** [comment]
Noted, got the below matches,

```shell
{"level":"warn","service":"blocks","impl":"standard","slot":33,"error":"failed to update attestations: failed to obtain database attestation: failed to fetch beacon committees: failed to obtain beacon committees","time":"2023-01-13T21:35:38Z","message":"Failed to update block"}
{"level":"trace","service":"blocks","impl":"standard","slot":5568477,"block_root":"0xd638a6372583029a44192fdebfbffc0699c47e622ad20e0291c9512d819eceec","state_root":"0x7c94e6cdf31c830debb0c67f353d15dac7ba0e3423f1132c14a85cef3f9b6db1","epoch_transition":false,"time":"2023-01-13T21:35:50Z","message":"Handler called"}
{"level":"trace","service":"blocks","impl":"standard","slot":33,"time":"2023-01-13T21:35:50Z","message":"Updating block for slot"}
{"level":"warn","service":"blocks","impl":"standard","slot":33,"error":"failed to update attestations: failed to obtain database attestation: failed to fetch beacon committees: failed to obtain beacon committees","time":"2023-01-13T21:35:50Z","message":"Failed to update block"}
{"level":"trace","service":"blocks","impl":"standard","slot":5568478,"block_root":"0x666359bf89d96393a6f9dec4747ad1959da6e8c840ac878daa9098fcce8f8ea3","state_root":"0xa1d6b083b5fec405b6ad579358bb7edf376e6b7063f6e2ab4edd5b10cc2c9a57","epoch_transition":false,"time":"2023-01-13T21:36:02Z","message":"Handler called"}
{"level":"trace","service":"blocks","impl":"standard","slot":33,"time":"2023-01-13T21:36:02Z","message":"Updating block for slot"}
```

**Comment by mcdee** [comment]
That shows that the blocks are waiting on the beacon committee information, so can you `grep '"service":"beaconcommittees"' chaind.log` to see why that service is not updating?

**Comment by sivathanuk** [comment]
Sure, just did. Here are the matches for `grep '"service":"beaconcommittees"' chaind.log`

```shell
{"level":"trace","service":"beaconcommittees","impl":"standard","epoch":0,"time":"2023-01-13T21:30:03Z","message":"Updating beacon committees"}
{"level":"trace","service":"beaconcommittees","impl":"standard","epoch":1,"time":"2023-01-13T21:30:03Z","message":"Updating beacon committees"}
{"level":"warn","service":"beaconcommittees","impl":"standard","epoch":1,"error":"failed to fetch beacon committees: failed to obtain beacon committees","time":"2023-01-13T21:30:03Z","message":"Failed to update beacon committees"}
{"level":"trace","service":"beaconcommittees","impl":"standard","epoch":0,"time":"2023-01-13T21:36:27Z","message":"Updating beacon committees"}
{"level":"trace","service":"beaconcommittees","impl":"standard","epoch":1,"time":"2023-01-13T21:36:27Z","message":"Updating beacon committees"}
{"level":"warn","service":"beaconcommittees","impl":"standard","epoch":1,"error":"failed to fetch beacon committees: failed to obtain beacon committees","time":"2023-01-13T21:36:27Z","message":"Failed to update beacon committees"}
```

**Comment by mcdee** [comment]
Okay, so you now need to look in the logs for the `failed to obtain beacon committees` line, and look above it to see what has failed.  If  you prefer, you can upload the logs somewhere and I can take a look.

**Comment by sivathanuk** [comment]
Yeah sure, I just did the grep ans shared the matches below. 

I can upload the logs in the cloud as well. The size of the log fie is 1.3 Gb. If it works, I can SCP the file to my local, upload it to Dropbox and share the Dropbox link to your public email [Jim@mcdee.net](mailto:Jim@mcdee.net). 

```shell
{"level":"warn","service":"blocks","impl":"standard","slot":33,"error":"failed to update attestations: failed to obtain database attestation: failed to fetch beacon committees: failed to obtain beacon committees","time":"2023-01-13T21:55:03Z","message":"Failed to update block"}
{"level":"warn","service":"blocks","impl":"standard","slot":33,"error":"failed to update attestations: failed to obtain database attestation: failed to fetch beacon committees: failed to obtain beacon committees","time":"2023-01-13T21:55:13Z","message":"Failed to update block"}
{"level":"warn","service":"blocks","impl":"standard","slot":33,"error":"failed to update attestations: failed to obtain database attestation: failed to fetch beacon committees: failed to obtain beacon committees","time":"2023-01-13T21:55:27Z","message":"Failed to update block"}
{"level":"warn","service":"beaconcommittees","impl":"standard","epoch":1,"error":"failed to fetch beacon committees: failed to obtain beacon committees","time":"2023-01-13T21:55:40Z","message":"Failed to update beacon committees"}
```

**Comment by sivathanuk** [comment]
Just shared the chaind.log file (~2 Gb) via Dropbox to [Jim@mcdee.net](mailto:Jim@mcdee.net). Please check and let me know if you're able to fetch the log file.

**Comment by mcdee** [comment]
The problem appears to be that the beacon node is not returning information about the comittees, which is causing everything else to be on hold.  Please could you run the following and paste the output?

```
time curl -s -o /dev/null -w "%{http_code}" http://localhost:5052/eth/v1/beacon/states/32/committees
```

This will show if the data is being returned, and if so how long it takes, outside of `chaind`.

**Comment by sivathanuk** [comment]
Just checked, got the below response.

```shell
404
real    0m0.010s
user    0m0.008s
sys     0m0.000s
```

Here is my beacon node (lighthouse) config, 
```YAML
sudo ./lighthouse bn --network mainnet --execution-endpoint http://localhost:8551 --datadir /data-mainnet --execution-jwt keystore/jwt-secret-key --checkpoint-sync-url https://mainnet.checkpoint.sigp.io --disable-deposit-contract-sync --metrics  --debug-level info  --eth1 --http --http-allow-sync-stalled --enr-udp-port=9000 --enr-tcp-port=9000 --discovery-port=9000
```
checked the directory info for `data-mainnet` dir (datadir for Lighthouse) and got the below output, 
```shell
947M    ./logs
125G    ./chain_db
379M    ./freezer_db
12K     ./network
126G    .
```

**Comment by mcdee** [comment]
That lighthouse configuration will nott be suitable for an archive node, `chaind` will not be able to use it.  You will need to:
- stop `chaind`
- remove and recreate the chain database
- stop lighthouse
- remove the lighthouse data-mainnet directory
- restart lighthouse with the additional options `--slots-per-restore-point=64` (or `32` if you have lots of disk space)

You then need to wait for lighthouse to fully sync.  With a `slots-per-restore-point` of 64 you should expect the `freezer_db` to be around 3TB, and double that if you use 32.  Only then should you restart `chaind`, and it should sync.

FYI I just tried a new instance of `chaind` from a lighthouse node synced as per above and it started to load fine, so this is very likely your issue.

**Comment by sivathanuk** [comment]
Thanks a lot for your timely support, checking and updating. I’ll look into the hardware, upgrade it and follow the above steps mentioned. 

Closing the issue, since it’s resolved now.

**Comment by sivathanuk** [comment]
Just an update I got from the Lighthouse team. Thought, I'l share it, 

TL;DR: Lighthouse is working on an alpha release of a database  that uses vastly less space (250 GB vs current 4TB+) 
for historic data. 

> if you’re planning to do a lot of historic analysis I would recommend increasing the amount of historic data stored on disk, see: https://lighthouse-book.sigmaprime.io/advanced_database.html. Currently this requires a lot of disk space (4TB+ for the fastest option). 

> We have a prototype that uses vastly less space (250GB) which I would like to cut an alpha release for ASAP (maybe as soon as next week). If you’re interested in that keep an eye on our announcements channel

**Comment by mcdee** [comment]
If you are running lighthouse with checkpoint sync you will need to also add the `--reconstruct-historical-states` flag or else you won't have a full set of data.

Also, if you only have 4TB of total disk space you will run out of space when populating the database, lighthouse will take up 3TB and the database will take another few TB on top.

**Comment by sivathanuk** [comment]
Thanks a lot for the suggestion. I have added the `--reconstruct-historical-states` flag to lighthouse config and re-running the sync. Just wondering if I can explore chaind by pushing only 3 months of eth data (by modifiying chaind config:   `--blocks. start-slot: 0`) to the database?

**Comment by mcdee** [comment]
`start-slot` is designed to allow rerunning of missed data rather than starting at a later point, it will likely cause failures elsewhere so I wouldn't recommend doing that.  You can always manually prune the data afterwards if required.

I can't say how well chaind would run on an hdd, I have never tried it.

**Comment by sivathanuk** [comment]
Got it. Thanks a lot for the quick response. 

Is there a way I can move the synced freezer_db data to a HDD and use the SSD to sync eth nodes and run chaind? Just trying to figure out how to use chaind with a 4TB SSD and a 8TB HDD. Kindly let me know if you have any suggestions. Link to resources should also help. Thanks.

**Comment by mcdee** [comment]
Sorry, you'd have to check with the documentation for the nodes that you are running for this information.

**Comment by sivathanuk** [comment]
Sure, thanks a lot. Already checking with them. Thanks again for your time and guidance.
